### DISCUSSION TITLE
Classifier tester

### DISCUSSION THREAD

**Discussion by gd-6** [KevinMusgrave/pytorch-metric-learning#716] [Q&A]
Hi,

I'm currently testing this library to achieve two main objectives:

**Training and testing a classifier:**
I would like to train a model to classify images, but without comparing the images between them. Specifically, I want to pass a single image to the model and predict its class.
I noticed the "train_with_classifier" example, which includes an additional "classifier" layer. However, from what I understand, most testers use the "embedder" for prediction and metric computation.
My questions are:

- Is there a way to calculate accuracy using the "classifier" layer?
- Alternatively, is it possible to make class predictions using only the "embedder", perhaps by comparing the trained embeddings?

**Hierarchical prediction:**
For the second task, I aim to implement hierarchical classification. For example, using the CIFAR100 dataset, I would like the network to learn the hierarchical structure of the dataset.
Are there any methods or loss functions implemented in the library that support hierarchical learning for tasks like this?

Thank you for maintaining this repository! I look forward to your guidance.

**Comment by KevinMusgrave** [comment]
> Is there a way to calculate accuracy using the "classifier" layer?

There's no built-in way to compute accuracy using the classifier layer. You'll have to loop through your evaluation dataset, and for each sample get the output of the classifier layer, and compare the max index of the output with the label of each sample.

> Alternatively, is it possible to make class predictions using only the "embedder", perhaps by comparing the trained embeddings?

This is probably the best approach: https://github.com/KevinMusgrave/pytorch-metric-learning/issues/583#issuecomment-1409847082

> Are there any methods or loss functions implemented in the library that support hierarchical learning for tasks like this?

No unfortunately there aren't any in this library. Here's a potentially relevant repo though: https://github.com/salesforce/hierarchicalContrastiveLearning
